src/docs/man/hdb.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HDB 1"
.TH HDB 1 "2006-03-28" "Tue Mar 28 16:25:00 EST 2006" "HDB Documenation"
.SH "NAME"
hdb \- multipurpose tool for accessing Hampus Database. 
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
hdb [\s-1OPTIONS\s0]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
hdb is a tool for accessing and modifying data in a Hampus Database. 
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fB\-l\fR <list>  list to operate on 
.PP
\&\fB\-c\fR <list>  create list 
.PP
\&\fB\-r\fR <list>  remove list 
.PP
\&\fB\-p\fR <list>  print all values from list 
.PP
\&\fB\-P\fR <list>  print all values from list with full path 
.PP
\&\fB\-s\fR <value> set value to list 
.PP
\&\fB\-g\fR <key>   get key from list 
.PP
\&\fB\-d\fR <key>   delete key from list
.PP
\&\fB\-k\fR <key>   key to use with \-s switch
.PP
\&\fB\-n\fR <num>   get record number from list (first = recno 1) 
.PP
\&\fB\-L\fR <list>  print list of lists
.PP
\&\fB\-S\fR <list>  print sublist of this list (one level) 
.PP
\&\fB\-Z\fR <list>  print sublist of this list with full path (one level)
.PP
\&\fB\-N\fR <list>  print number of elements in this sublist list (one level) 
.PP
\&\fB\-x\fR <list>  dump data in xml like format 
.PP
\&\fB\-f\fR <list>  dump data in flat file format (list/sublist/key=value)
.PP
\&\fB\-m\fR <list>  use meta values 
.PP
\&\fB\-t\fR <type>  use meta type (str,link,exe,fil,int). str=default
.PP
\&\fB\-b\fR         output in bash friendly format
.PP
\&\fB\-D\fR         output debug info
.PP
\&\fB\-w\fR         wipe all lists
.PP
\&\fB\-b\fR         quite output
.PP
\&\fB\-R\fR <\s-1DIR\s0>   use \s-1DIR\s0 as hdb root
.PP
\&\fB\-h\fR         show help
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Fetch a key
        hdb \-l list/sublist \-g key
.PP
Set a value
	hdb \-l list/sublist \-k key \-s value
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
hdb-query returns a zero  exist  status if it succeeds
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
\&\s-1HDB_ROOT\s0 
        A directory path to override the default database directory. Default root is /var/db/hdb
.SH "AUTHOR"
.IX Header "AUTHOR"
Hampus Sodestrom hampus@sxp.se
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIhdb\-get\fR\|(1) \fIhdb\-del\fR\|(1) \fIhdb\-cut\fR\|(1) \fIhdb\-cut\fR\|(1) \fIhdb\-dump\fR\|(1) \fIhdb\-import\fR\|(1) \fIhdb\-filter\fR\|(1) \fIhdb\-find\fR\|(1) \fIhdb\-edit\fR\|(1) 


src/perl/HDB.pm
#!/usr/bin/perl -w
#########################################################################
# HDB, Copyright (C) 2006 Hampus Soderstrom
# All rights reserved.  Email: hampus@sxp.se
#
# This library is free software; you can redistribute it and/or         
# modify it under the terms of The GNU Lesser General Public License as
# published by the Free Software Foundation; either version 2.1 of the 
# License, or (at your option) any later version. The text of the GNU
# Lesser General Public License is included with this library in the
# file LICENSE.TXT.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the files
# LICENSE.TXT for more details.
#########################################################################
package HDB;
use IO::Socket;
use URI::Escape;
use strict;

my ($hdb_error) = 'ENONE';

sub new 
{
    my ($class, %h) = @_;
    my ($host, $port);
    my $self = {};    
    bless $self, $class;

    $self->set_error;
    $self->{'quit'} = 0;

    if (defined $h{'Host'})
    {
	($host, $port) = split(/:/, $h{'Host'});
    }
    $host = "127.0.0.1" unless $host;
    $port = "12429" unless $port;
    $self->{'host'} = $host;
    $self->{'port'} = $port;
    $self->{'doAutoflush'} = 0;

    $self->connect() || return undef;
    $self;
}

sub autoflush
{
    my ($self, $doAutoflush) = @_;
    if ($doAutoflush == 0)
    {
	$self->{'doAutoflush'} = 0;
    }
    else
    {
	$self->{'doAutoflush'} = 1;	
    }
}

sub connect
{
    my ($self) = @_;
    $self->{'sock'} = new IO::Socket::INET( PeerAddr => $self->{'host'},
					    PeerPort => $self->{'port'},
					    Type => SOCK_STREAM,
					    ReuseAddr => 1,
					    Proto => 'tcp',
					    TimeOut => 3
					    ) or return $self->set_error('ESOCKETFAIL');
    $self->{'sock'}->getline();
    if (defined $self->{'root'})
    {
	$self->ROOT($self->{'root'}, 1);
    }
    else
    {
	$self->{'root'} = $self->ROOT();
    }
}

sub DESTROY
{
     my ($self) = @_;
     $self->QUIT;
}

sub QUIT
{
     my ($self) = @_;
     if ($self->{'quit'} == 0)
     { 
	 $self->execute("QUIT");
	 $self->{'sock'}->shutdown(2);
	 $self->{'quit'} = 1;
     }
}

sub execute
{
    my ($self, $command, $retries) = @_;
    $retries = 0 unless $retries;
    return 0 if ($retries > 3);
    $self->FLUSH() if ($self->{'doAutoflush'} == 1);
    if (defined $self->{'sock'})
    {
	$self->{'sock'}->write("$command\n");
    }
    else
    {
	# Server dead ?!?
	# Reconnect and execute command again
	sleep 1;
	$self->connect || return 0;
	return $self->execute($command, ++$retries);
    }
    @{$self->{'result'}} = ();
    my $line = $self->{'sock'}->getline() || "#DEAD";
    chomp $line;
    while ($line !~ /^\#(.*)/)
    {
	push @{$self->{'result'}}, $line;
	$line = $self->{'sock'}->getline() || "#DEAD";
	chomp $line;
    }
    $line =~ /\#(.*)/;
    $self->{'status'} = $1;
    if ($1 eq "OK")
    {
	return 1;
    }
    elsif ($1 eq "DEAD")
    {
	# Server dead ?!?
	# Reconnect and execute command again
	sleep 1;
	$self->connect ||return 0;
	return $self->execute($command, ++$retries);
    }
    else
    {
	return 0;
    }
}

sub GET
{
    my ($self, $list, $val) = @_; 
    #$self->execute("GET $list $val");
    $self->execute("0 $list $val");
    return ${$self->{'result'}}[0];
}

sub UPDATE
{
    my ($self, $list, $key, $val) = @_;
    $val =~ s/\\/\\\\/g;
    $val =~ s/\"/\\\"/g;
    return $self->execute("UPDATE $list $key " . uri_escape($val));
}

sub SET
{
    my ($self, $list, $key, $val) = @_;
    $val =~ s/\\/\\\\/g;
    $val =~ s/\"/\\\"/g;
    return $self->execute("1 $list $key " . uri_escape($val));
}

sub SYNC 
{
    my ($self) = @_;
    #return $self->execute("SYNC");
    return $self->execute("23");
}

sub LIST
{
    my ($self) = @_;
    #$self->execute("LIST");
    $self->execute("21");
    return @{$self->{'result'}};
}

sub SUBLIST
{
    my ($self, $list) = @_;
    #$self->execute("SUBLIST $list");
    $self->execute("30 $list");
    return @{$self->{'result'}};
}

sub PRINT
{
    my ($self, $list) = @_;
    my %h = ();
    #$self->execute("PRINT $list");
    $self->execute("2 $list");
    for (@{$self->{'result'}})
    {
	my ($key, $val) = split;
	$h{$key} = $val;
    }
    return %h;
}

sub ROOT
{
    my ($self, $root, $force) = @_;
    if (!defined($root))
    {
	#$self->execute("ROOT");
	$self->execute("33");
	return ${$self->{'result'}}[0];
    }
    elsif ($root ne $self->{'root'} || $force == 1)
    {
	$self->{'root'} = $root;
	#return $self->execute("ROOT $root");
	return $self->execute("33 $root");
    }
    else
    {
	return $self->{'root'};
    }
}

sub DUMP
{
    my ($self, $list) = @_;
    #$self->execute("DUMP $list");
    $self->execute("3 $list");
    return @{$self->{'result'}};
}

sub set_error 
{
    my ($self, $error) = @_;
    $hdb_error = $self->{'error'} = defined $error ? $error : 'ENONE';
    return undef;
}

sub get_error 
{
    my ($self) = @_;
    $self->{'error'};
}

sub strerror 
{
    my ($self, $error) = @_;

    my %errors = (
		  'ENONE',      'none',
		  'ESELECTFAIL','select creation failed',
		  'ETIMEOUT',   'timed out waiting for packet',
		  'ESOCKETFAIL','socket creation failed',
		  'ENOHOST',    'no host specified',
		  'EBADAUTH',   'bad response authenticator',
		  'ESENDFAIL',	'send failed',
		  'ERECVFAIL',	'receive failed'
		  );

    return $errors{$hdb_error} unless ref($self);
    $errors{defined $error ? $error : $self->{'error'}};
}

1;
